Tests/NostrSDKTests/NIP44v2EncryptingTests.swift

```diff
@@ -0,0 +1,191 @@
+            XCTAssertEqual(messageKeys.chaChaNonce.hexString, vector.chaChaNonce)
+            XCTAssertEqual(messageKeys.hmacKey.hexString, vector.hmacKey)
+        }
+    }
+
+    /// Take unpadded length (first value), calculate padded length (second value).
+    func testValidCalculatePaddedLength() throws {
+        let calculatePaddedLengthVectors = try XCTUnwrap(vectors.v2.valid.calculatePaddedLength)
+        try calculatePaddedLengthVectors.forEach { vector in
+            XCTAssertEqual(vector.count, 2)
+            let paddedLength = try calculatePaddedLength(vector[0])
+            XCTAssertEqual(paddedLength, vector[1])
+        }
+    }
+
+    /// Emulate real conversation with a hardcoded nonce.
+    /// Calculate pub2 from sec2, verify conversation key from (sec1, pub2), encrypt, verify payload.
+    /// Then calculate pub1 from sec1, verify conversation key from (sec2, pub1), decrypt, verify plaintext.
+    func testValidEncryptDecrypt() throws {
+        let encryptDecryptVectors = try XCTUnwrap(vectors.v2.valid.encryptDecrypt)
+        try encryptDecryptVectors.forEach { vector in
+            let sec1 = vector.sec1
+            let sec2 = vector.sec2
+            let expectedConversationKey = vector.conversationKey
+            let nonce = try XCTUnwrap(vector.nonce.hexDecoded)
+            let plaintext = vector.plaintext
+            let payload = vector.payload
+
+            let keypair1 = try XCTUnwrap(Keypair(hex: sec1))
+            let keypair2 = try XCTUnwrap(Keypair(hex: sec2))
+
+            // Conversation key from sec1 and pub2.
+            let conversationKey1Bytes = try conversationKey(
+                privateKeyA: keypair1.privateKey,
+                publicKeyB: keypair2.publicKey
+            ).bytes
+            XCTAssertEqual(expectedConversationKey, Data(conversationKey1Bytes).hexString)
+
+            // Verify payload.
+            let ciphertext = try encrypt(
+                plaintext: plaintext,
+                conversationKey: conversationKey1Bytes,
+                nonce: nonce
+            )
+            XCTAssertEqual(payload, ciphertext)
+
+            // Conversation key from sec2 and pub1.
+            let conversationKey2Bytes = try conversationKey(
+                privateKeyA: keypair2.privateKey,
+                publicKeyB: keypair1.publicKey
+            ).bytes
+            XCTAssertEqual(expectedConversationKey, Data(conversationKey2Bytes).hexString)
+
+            // Verify that decrypted data equals the plaintext that we started off with.
+            let decrypted = try decrypt(payload: payload, conversationKey: conversationKey2Bytes)
+            XCTAssertEqual(decrypted, plaintext)
+        }
+    }
+
+    /// Emulate real conversation with only the public encrypt and decrypt functions,
+    /// where the nonce used for encryption is a cryptographically secure pseudorandom generated series of bytes.
+    func testValidEncryptDecryptRandomNonce() throws {
```

Comment:
Added this function so that it's a true test of how the API would be used in the wild.